File: src/test/java/nihilite/test/CapturingLogHandler.java
package nihilite.test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.LogRecord;

public final class CapturingLogHandler extends Handler {

    private final List<LogRecord> captured;

    public CapturingLogHandler() {
        this.captured = Collections.synchronizedList(new ArrayList<>());
    }

    public List<LogRecord> captured() {
        return captured;
    }

    @Override
    public void publish(LogRecord record) {
        if (record.getLevel().intValue() >= Level.WARNING.intValue()) {
            captured.add(record);
        }
    }

    @Override
    public void flush() {}

    @Override
    public void close() {}
}

File: src/test/clojure/nihilite/test/capturing_log_handler.clj
(ns nihilite.test.capturing-log-handler
  "java.util.logging.Handler implementation that captures all log records
   published at Level >= WARNING. Replaces the previous Java test class
   of the same name.

   Returns a plain Clojure map with `:handler` (the proxy implementing
   `java.util.logging.Handler`) and `:captured` (synchronized list of
   LogRecord instances)."
  (:import [java.util.logging Handler Level LogRecord]
           [java.util ArrayList Collections]))

(defn- capturing-handler
  "Returns a proxy implementing java.util.logging.Handler; the captured
   list is mutated in-place on each `publish` call."
  [captured]
  (proxy [Handler] []
    (publish [record]
      (when (>= (.intValue (.getLevel ^LogRecord record)) (.intValue Level/WARNING))
        (.add ^java.util.List captured record)))
    (flush [])
    (close [])))

(defn make
  "Returns a map with `:handler` (the proxy) and `:captured` (the
   synchronized list of LogRecord instances)."
  []
  (let [captured (Collections/synchronizedList (ArrayList.))]
    {:handler (capturing-handler captured)
     :captured captured}))

(defn captured
  "Returns the captured records list from a handler map created by `make`."
  [m]
  (when (map? m) (:captured m)))
